verification/cbmc/sources/make_common_data_structures.c

```diff
@@ -126,21 +127,55 @@ bool aws_cryptosdk_edk_list_elements_are_bounded(const struct aws_array_list *co
-        __CPROVER_assume(list->data == NULL);
-        list->alloc = can_fail_allocator();
-    } else {
-        size_t max_length = list->current_size / sizeof(struct aws_cryptosdk_edk);
-        list->data        = bounded_malloc(sizeof(struct aws_cryptosdk_edk) * max_length);
-        list->alloc       = nondet_bool() ? NULL : can_fail_allocator();
+    if (list != NULL) {
+        if (list->current_size == 0) {
+            __CPROVER_assume(list->data == NULL);
+            list->alloc = can_fail_allocator();
+        } else {
+            size_t max_length = list->current_size / sizeof(struct aws_cryptosdk_edk);
+            list->data        = bounded_malloc(sizeof(struct aws_cryptosdk_edk) * max_length);
+            list->alloc       = nondet_bool() ? NULL : can_fail_allocator();
+        }
     }
 }
 
 void ensure_cryptosdk_edk_list_has_allocated_list_elements(struct aws_array_list *list) {
-    for (size_t i = 0; i < list->length; ++i) {
-        struct aws_cryptosdk_edk *data = (struct aws_cryptosdk_edk *)list->data;
-        ensure_cryptosdk_edk_has_allocated_members(&(data[i]));
+    if (aws_cryptosdk_edk_list_is_valid(list)) {
+        for (size_t i = 0; i < list->length; ++i) {
+            struct aws_cryptosdk_edk *data = (struct aws_cryptosdk_edk *)list->data;
+            ensure_cryptosdk_edk_has_allocated_members(&(data[i]));
+        }
+    }
+}
+
+struct aws_cryptosdk_hdr *ensure_nondet_hdr_has_allocated_members(const size_t max_table_size) {
+    struct aws_cryptosdk_hdr *hdr = malloc(sizeof(*hdr));
+    if (hdr != NULL) {
+        hdr->alloc     = nondet_bool() ? NULL : can_fail_allocator();
+        hdr->frame_len = malloc(sizeof(hdr->frame_len));
+        ensure_byte_buf_has_allocated_buffer_member(&hdr->iv);
+        ensure_byte_buf_has_allocated_buffer_member(&hdr->auth_tag);
+        ensure_byte_buf_has_allocated_buffer_member(&hdr->message_id);
+        ensure_byte_buf_has_allocated_buffer_member(&hdr->alg_suite_data);
+        ensure_allocated_hash_table(&hdr->enc_ctx, max_table_size);
+        ensure_cryptosdk_edk_list_has_allocated_list(&hdr->edk_list);
+    }
+    return hdr;
+}
+
+bool aws_cryptosdk_hdr_members_are_bounded(
+    const struct aws_cryptosdk_hdr *hdr, const size_t max_edk_item_size, const size_t max_item_size) {
+    if (hdr != NULL) {
+        /*IV buffer length might need further constraints, this is done in the harness when necessary */
+        return aws_cryptosdk_edk_list_is_bounded(&hdr->edk_list, max_edk_item_size) &&
+               (!aws_cryptosdk_edk_list_is_valid(&hdr->edk_list) ||
+                aws_cryptosdk_edk_list_elements_are_bounded(&hdr->edk_list, max_item_size)) &&
+               aws_byte_buf_is_bounded(&hdr->message_id, max_item_size) &&
+               aws_byte_buf_is_bounded(&hdr->iv, max_item_size) &&
+               aws_byte_buf_is_bounded(&hdr->alg_suite_data, max_item_size) &&
+               aws_byte_buf_is_bounded(&hdr->auth_tag, max_item_size);
```

Comment:
nitpick: It'll be easier to maintain these `byte_buf`-related lines if they have the same order whenever they appear, e.g. in order of declaration within the header struct

Reply:
Great idea! Updated to use the same order as in the header struct